crates/api/src/op_store.rs

```diff
@@ -0,0 +1,86 @@
+//! Kitsune2 op store types.
+
+use crate::{K2Result, OpId, Timestamp};
+use futures::future::BoxFuture;
+use std::cmp::Ordering;
+use std::sync::Arc;
+
+/// An op with metadata.
+///
+/// This is the basic unit of data in the kitsune2 system.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct MetaOp {
+    /// The id of the op.
+    pub op_id: OpId,
+
+    /// The creation timestamp of the op.
+    ///
+    /// This must be the same for everyone who sees this op.
+    ///
+    /// The host must reject the op if the timestamp does not agree with any timestamps inside the
+    /// op data.
+    pub timestamp: Timestamp,
+
+    /// The actual op data.
+    pub op_data: Vec<u8>,
+
+    /// Crdt-style add-only opaque implementor-use flags.
+    pub op_flags: std::collections::HashSet<String>,
+}
+
+impl Ord for MetaOp {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (&self.timestamp, &self.op_id).cmp(&(&other.timestamp, &other.op_id))
+    }
+}
+
+impl PartialOrd for MetaOp {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// The API that a kitsune2 host must implement to provide data persistence for kitsune2.
+pub trait OpStore: 'static + Send + Sync + std::fmt::Debug {
```

Comment:
This is definitely not finished. We need to be able to partition by hash range too. Not sure what granularity we want the "space" at. Whether we create one op store per space or it needs to be passed through the API. I'm thinking the former but I'm not planning on dealing with those things as part of this change.

Reply:
Not sure I understand the "partition by hash range" comment. The pagination for a catch-up sync would just be by timestamp cursor, and the only other query mode we need is for an arc-range, right?

I agree we should have a different op store instance per space (though holochain may or may not choose to use the same database for all the instances).

Reply:
I think it should be by timestamps cursor, per hash range. So initially as we're doing full or empty arc, it's only full arc where we fetch ops and so yes it's just a time cursor. When we have the hash space partitioned, I'm thinking at the moment that hash partitioning lives at the top level and then we slice time within that hash partition.

So it looks like concentric circles, cut up like a pizza. 

For that to work, looking up data by time slice is always "what is the next block of ops from time 203942380492834, limited to 5Mb, in the hash range 0x...0x"

Reply:
Okay, by hash range, you actually mean arc range, or hash location range. (just to distinguish from holochain's hash-bounded-queries).